Feat: added caching schema for debuginfo type names

Sl1mb0 opened the pull request:
This is a WIP addressing #86431. This PR aims to accomplish two things:

- [x] Implements a caching schema to avoid redundant calls to `compute_debuginfo_type_name()`.
- [ ] Add profiling support for determining the amount of redundant computations that _would_ be performed, if the caching schema were not implemented. 
    - As for how: I plan to just use counters to track how many times we call `compute_debuginfo_type_name()` and how many times we return a name for the cache. @rylev also suggested that determining how many individual memory allocations occur each time redundant work is performed, but I am unsure of how to go about this.

rust-highfive commented:
r? @jackh726

(rust-highfive has picked a reviewer for you, use r? to override)

Sl1mb0 commented:
r? @michaelwoerister

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3581147954) [(plain)](https://github.com/rust-lang/rust/commit/16433a9320b4f2ef8f036e06c0756d7f1b7bda5c/checks/3581147954/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs at line 162:
     fn register_unique_id_with_type_name(
         &mut self,
         unique_type_id: UniqueTypeId,
+        type_name: String,
     ) {
     ) {
         if self.unique_id_to_type_name.insert(unique_type_id, type_name).is_some() {
             bug!(
Diff in /checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs at line 350:
 }
 
-fn check_type_name_cache(
-fn check_type_name_cache(
-    cx: &CodegenCx<'ll, 'tcx>,
-    qualified: bool,
-) -> String {
-) -> String {
+fn check_type_name_cache(cx: &CodegenCx<'ll, 'tcx>, ty: Ty<'tcx>, qualified: bool) -> String {
     let mut type_map = debug_context(cx).type_map.borrow_mut();
     let unique_type_id = type_map.get_unique_type_id_of_type(cx, ty);
     match type_map.find_type_name_for_unique_id(unique_type_id) {
Diff in /checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs at line 361:
-        Some(type_name) => { type_name },
+        Some(type_name) => type_name,
         None => {
             let type_name = compute_debuginfo_type_name(cx.tcx, ty, qualified);
-            type_map.register_unique_id_with_type_name(
-                unique_type_id,
-                type_name.clone(),
-            );
+            type_map.register_unique_id_with_type_name(unique_type_id, type_name.clone());
-        },
+        }
     }
 }
 }
 
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/namespace.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/intrinsic.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/backend.rs" "/checkout/compiler/rustc_codegen_ssa/src/traits/coverageinfo.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/mod.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

Sl1mb0 commented:
FWIW, here are some interesting statistics from profiling the nightly and my stage2 compiler on the `regex` crate and comparing the results:

- Total cpu time: +799.441527ms increase
- time spent in `compute_debuginfo_type_name()`
    - nightly: `| 23.44ms   | 0.242           | 25.23ms  | 25723      |`
    - stage2: `| 25.78ms   | 0.253           | 26.41ms  | 25723      |`

I would love some ideas for how I can more accurately profile this code, as the above results don't make sense to me. 

If we are caching all names, allowing us to lookup repeat ones instead of recomputing them; shouldn't there be less calls to `compute_debuginfo_type_name()` in general, and thus less time spent in `compute_debuginfo_type_name()`?

bjorn3 commented on compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs:
```diff
@@ -106,6 +106,8 @@ pub struct TypeMap<'ll, 'tcx> {
     /// A map from types to `UniqueTypeId`. This is an N:1 mapping.
     type_to_unique_id: FxHashMap<Ty<'tcx>, UniqueTypeId>,
+    /// A map from `UniqueTypeId` to it's type-name string.
+    unique_id_to_type_name: FxHashMap<UniqueTypeId, String>,
```
A `UniqueTypeId` already contains the string. A `Symbol` is an interned string.

Sl1mb0 replied:
So we can just check

bjorn3 replied:
Right, seems like it is a hex encoded hash.

lqd commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 03361c46dd5c34a4b73ec159438486c64c15310d with merge db7e55346581a97b2acf2430c0b7b42c9da07e0b...
<!-- homu: {"type":"TryBuildStarted","head_sha":"03361c46dd5c34a4b73ec159438486c64c15310d","merge_sha":"db7e55346581a97b2acf2430c0b7b42c9da07e0b"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3590934463)
Build commit: db7e55346581a97b2acf2430c0b7b42c9da07e0b (`db7e55346581a97b2acf2430c0b7b42c9da07e0b`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3590934463"},"merge_sha":"db7e55346581a97b2acf2430c0b7b42c9da07e0b"} -->

rust-timer commented:
Queued db7e55346581a97b2acf2430c0b7b42c9da07e0b with parent b0ee4951f0289aa31e54b101c5706f220315e197, future [comparison URL](https://perf.rust-lang.org/compare.html?start=b0ee4951f0289aa31e54b101c5706f220315e197&end=db7e55346581a97b2acf2430c0b7b42c9da07e0b).

rust-timer commented:
Finished benchmarking commit (db7e55346581a97b2acf2430c0b7b42c9da07e0b): [comparison url](https://perf.rust-lang.org/compare.html?start=b0ee4951f0289aa31e54b101c5706f220315e197&end=db7e55346581a97b2acf2430c0b7b42c9da07e0b).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf. 

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

michaelwoerister commented:
Thanks for the PR, @Sl1mb0! 

I think for this to really make a performance difference the caching would have to be integrated into `compute_debuginfo_type_name()` itself, so that two separate calls for two types like `foo::Bar1<foo::Baz>` and `foo::Bar2<foo::Baz>` would be able to re-use `foo::Baz`. You can experiment with adding such a cache as a parameter to `compute_debuginfo_type_name()`. It would be something like a `&mut FxHashMap<(Ty<'tcx>, bool), String>` where the `bool` in the key stands for the `qualified` parameter.

Sl1mb0 commented:
@michaelwoerister Right now I am implementing your suggestion, and really my only question is:

Where would you suggest the _actual_ caching occur? The only alternative I see to caching the result in`compute_debuginfo_type_name()` after we have called `push_debuginfo_type_name()` is within `push_debuginfo_type_name()`, and that seems like it will add some complexity unless it occurs at the end of the call, in which case it isn't much different than the current implementation.

(this is a dumb question, I am still asking it)

As for checking the cache; I don't see the difference between checking before the call to `compute_debuginfo_type_name()` and checking right away within the call itself; unless of course there is something I don't know :)

Sl1mb0 commented:
@michaelwoerister Do you have any suggestions on where caching should occur?

michaelwoerister commented:
The cache could live in the `CrateDebugContext`, I think.

https://github.com/rust-lang/rust/blob/e2750baf53aaa60db95f10759f6cf9463dc5a6bd/compiler/rustc_codegen_llvm/src/debuginfo/mod.rs#L59-L72

You can try to add a `type_name_cache: RefCell<FxHashMap<(Ty<'tcx>, bool), String>>` there and see if you run into any problems.

Sl1mb0 commented:
Sorry I should have been more clear. 

In this PR I've stored the cache in the `TypeMap`, but should it be moved? 

And to clarify, where should the call to `FxHashMap::insert()` be happening within `push_debuginfo_type_name()`? I know that ultimately we should be checking the cache before any recursive calls, but I've been running into problems when trying to insert into the cache, and am a little lost.

bors commented:
:umbrella: The latest upstream changes (presumably #88956) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Sl1mb0 commented:
So I've changed a few things; but now the compiler is panicking due to a `BorrowMut` error, and I am unsure why.

My guess is that it's because I am passing around a mutable reference to the cache, as opposed to passing around a `RefCell`, and attempting to borrow from that `RefCell`.

The changes in these commits are:

- [x] Moved `type_name_cache` up to the `CrateDebugContext`
- [x] Add `type_name_cache` arg to `compute_debuginfo_type_name()` and it's helper functions.
- [x] Upon entry into `push_debuginfo_type_name()` check if we have computed this name before, compute it if we  haven't, and return the computed name.

Sl1mb0 commented:
I changed `&mut` to `&RefCell`, and everything compiles. However `debuginfo/pretty-slices.rs` fails when running `./x.py test debuginfo`.

`stdout`: 
```
failures:

---- [debuginfo-lldb] debuginfo/pretty-slices.rs stdout ----
NOTE: compiletest thinks it is using LLDB version 1205
NOTE: compiletest thinks it is using LLDB without native rust support

error: line not found in debugger output:  (&mut [i32]) $1 = size=4 { [0] = 2 [1] = 3 [2] = 5 [3] = 7 }
status: exit status: 0
command: "/usr/bin/python3" "/Users/t1mb0sl1c3/Developer/rust-lang/rust/src/etc/lldb_batchmode.py" "/Users/t1mb0sl1c3/Developer/rust-lang/rust/compiler/build/x86_64-apple-darwin/test/debuginfo/pretty-slices.lldb/a" "/Users/t1mb0sl1c3/Developer/rust-lang/rust/compiler/build/x86_64-apple-darwin/test/debuginfo/pretty-slices.lldb/pretty-slices.debugger.script"
stdout:
------------------------------------------
LLDB batch-mode script
----------------------
Debugger commands script is '/Users/t1mb0sl1c3/Developer/rust-lang/rust/compiler/build/x86_64-apple-darwin/test/debuginfo/pretty-slices.lldb/pretty-slices.debugger.script'.
Target executable is '/Users/t1mb0sl1c3/Developer/rust-lang/rust/compiler/build/x86_64-apple-darwin/test/debuginfo/pretty-slices.lldb/a'.
Current working directory is '/Users/t1mb0sl1c3/Developer/rust-lang/rust/compiler'
Creating a target for '/Users/t1mb0sl1c3/Developer/rust-lang/rust/compiler/build/x86_64-apple-darwin/test/debuginfo/pretty-slices.lldb/a'
settings set auto-confirm true

version
lldb-1205.0.27.3 Apple Swift version 5.4 (swiftlang-1205.0.26.9 clang-1205.0.19.55) 
command script import /Users/t1mb0sl1c3/Developer/rust-lang/rust/./src/etc/lldb_lookup.py
type synthetic add -l lldb_lookup.synthetic_lookup -x '.*' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)String$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^&(mut )?str$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^&(mut )?\[.+\]$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(std::ffi::([a-z_]+::)+)OsString$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)Vec<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)VecDeque<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)BTreeSet<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)BTreeMap<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(std::collections::([a-z_]+::)+)HashMap<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(std::collections::([a-z_]+::)+)HashSet<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)Rc<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(alloc::([a-z_]+::)+)Arc<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(core::([a-z_]+::)+)Cell<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(core::([a-z_]+::)+)Ref<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(core::([a-z_]+::)+)RefMut<.+>$' --category Rust
type summary add -F lldb_lookup.summary_lookup  -e -x -h '^(core::([a-z_]+::)+)RefCell<.+>$' --category Rust
type category enable Rust

breakpoint set --file 'pretty-slices.rs' --line 45
DEBUG: breakpoint added, id = 1
Breakpoint 1: where = a`pretty_slices::main::h926e671f9dfe1c03 + 155 at pretty-slices.rs:45:5, address = 0x0000000100002cab 
DEBUG: registering breakpoint callback, id = 1
Error while trying to register breakpoint callback, id = 1, message = error: could not get num args: can't find callable: breakpoint_callback

run
Process 39336 stopped * thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1 frame #0: 0x0000000100002cab a`pretty_slices::main::h926e671f9dfe1c03 at pretty-slices.rs:45:5 42 let mut mut_str_slice_buffer = String::from("mutable string slice"); 43 let mut_str_slice: &mut str = mut_str_slice_buffer.as_mut_str(); 44 -> 45 b(); // #break ^ 46 } Target 0: (a) stopped. Process 39336 launched: '/Users/t1mb0sl1c3/Developer/rust-lang/rust/compiler/build/x86_64-apple-darwin/test/debuginfo/pretty-slices.lldb/a' (x86_64) 
print slice
(&[i32]) $0 = size=3 { [0] = 0 [1] = 1 [2] = 2 } 
print mut_slice
(&mut &[i32]) $1 = { data_ptr = 0x00007ffeefbff288 length = 4 } 
print str_slice
(&str) $2 = "string slice" { data_ptr = 0x0000000100003e58 "string slicemutable string slicecalled `Result::unwrap()` on an `Err` value" length = 12 } 
print mut_str_slice
(&mut &str) $3 = { data_ptr = 0x0000000100606070 "mutable string slice" length = 20 } 
quit


------------------------------------------
stderr:
------------------------------------------

------------------------------------------



failures:
    [debuginfo-lldb] debuginfo/pretty-slices.rs

test result: FAILED. 1 passed; 1 failed; 131 ignored; 0 measured; 0 filtered out; finished in 4.69s

Some tests failed in compiletest suite=debuginfo mode=debuginfo host=x86_64-apple-darwin target=x86_64-apple-darwin
```

michaelwoerister commented:
Hm, that looks like a legitimate error: `(&mut [i32])` vs `(&mut &[i32])`. I'll try to take a look tomorrow. It looks like you are on the right track though otherwise.

Meanwhile, if you clean up the merge conflicts we can run some performance tests via CI.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-10`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3677312035) [(plain)](https://github.com/rust-lang/rust/commit/0d2f0bb311a1de15615c0de36f7ac9064f4bb7c5/checks/3677312035/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Suite("src/test/debuginfo") not skipped for "bootstrap::test::Debuginfo" -- not in ["src/tools/tidy"]
Check compiletest suite=debuginfo mode=debuginfo (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)

running 134 tests
iiiiiiiii..i.ii..i..i.iii..ii....i.i....ii...F......iiiii....ii.......ii......i....i.......FF.ii..i. 100/134
Some tests failed in compiletest suite=debuginfo mode=debuginfo host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
ii......iiii.....ii.........F...F.

---- [debuginfo-gdb] debuginfo/generic-struct.rs stdout ----
---- [debuginfo-gdb] debuginfo/generic-struct.rs stdout ----
NOTE: compiletest thinks it is using GDB with native rust support
NOTE: compiletest thinks it is using GDB version 8001001

error: line not found in debugger output: $1 = generic_struct::AGenericStruct<i32, i32> {key: 0, value: 1}
status: exit status: 0
command: "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/generic-struct.gdb/generic-struct.debugger.script"
------------------------------------------
GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x868: file /checkout/src/test/debuginfo/generic-struct.rs, line 82.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, generic_struct::main () at /checkout/src/test/debuginfo/generic-struct.rs:82
82     zzz(); // #break
$1 = generic_struct::AGenericStruct<i32, AGenericStruct<i32> {key: 0, value: 1}
$2 = generic_struct::AGenericStruct<AGenericStruct<i32, f64> {key: 2, value: 3.5}
$3 = generic_struct::AGenericStruct<AGenericStruct<AGenericStruct<i32, f64, AGenericStruct<i32> {key: 4.5, value: 5}
$4 = generic_struct::AGenericStruct<AGenericStruct<AGenericStruct<i32, f64, generic_struct::AGenericStruct<AGenericStruct<i32, AGenericStruct<AGenericStruct<i32, f64>> {key: 6.5, value: generic_struct::AGenericStruct<AGenericStruct<i32, f64> {key: 7, value: 8.5}}
A debugging session is active.

 Inferior 1 [process 19874] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr:
------------------------------------------


------------------------------------------


---- [debuginfo-gdb] debuginfo/pretty-slices.rs stdout ----
NOTE: compiletest thinks it is using GDB with native rust support
NOTE: compiletest thinks it is using GDB version 8001001

error: line not found in debugger output:  $2 = &mut [i32](size=4) = {2, 3, 5, 7}
status: exit status: 0
command: "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-slices.gdb/pretty-slices.debugger.script"
------------------------------------------
GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x31b8: file /checkout/src/test/debuginfo/pretty-slices.rs, line 45.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, pretty_slices::main () at /checkout/src/test/debuginfo/pretty-slices.rs:45
45     b(); // #break
$1 = &[i32](size=3) = {0, 1, 2}
$2 = &mut &[i32] {data_ptr: 0x7fffffffe060, length: 4}
$3 = "string slice"
$4 = &mut &str {data_ptr: 0x55555575aae0 "mutable string slice\000", length: 20}
A debugging session is active.

 Inferior 1 [process 20671] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr:
------------------------------------------


------------------------------------------


---- [debuginfo-gdb] debuginfo/pretty-huge-vec.rs stdout ----
NOTE: compiletest thinks it is using GDB with native rust support
NOTE: compiletest thinks it is using GDB version 8001001

error: line not found in debugger output: $2 = &[u8](size=1000000000) = {[...]...}
status: exit status: 0
command: "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-huge-vec.gdb/pretty-huge-vec.debugger.script"
------------------------------------------
GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x322a: file /checkout/src/test/debuginfo/pretty-huge-vec.rs, line 27.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, pretty_huge_vec::main () at /checkout/src/test/debuginfo/pretty-huge-vec.rs:27
27     zzz(); // #break
$1 = Vec(size=1000000000) = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...}
$2 = &[null<u8](size=1000000000) = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0...}
A debugging session is active.

 Inferior 1 [process 20672] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr:
------------------------------------------


------------------------------------------


---- [debuginfo-gdb] debuginfo/unsized.rs stdout ----
NOTE: compiletest thinks it is using GDB with native rust support
NOTE: compiletest thinks it is using GDB version 8001001

error: line not found in debugger output: $1 = unsized::Foo<[u8]> {value: [...]}
status: exit status: 0
command: "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/unsized.gdb/unsized.debugger.script"
------------------------------------------
GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x8c2: file /checkout/src/test/debuginfo/unsized.rs, line 32.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, unsized::main () at /checkout/src/test/debuginfo/unsized.rs:32
32     zzz(); // #break
$1 = unsized::Foo<[Foo<unsized::Foo<[u8]> {value: 0x7fffffffe0b8}
$2 = unsized::Foo<unsized::Foo<Foo<[Foo<unsized::Foo<[u8]>> {value: unsized::Foo<[Foo<unsized::Foo<[u8]> {value: 0x7fffffffe0b8}}
A debugging session is active.

 Inferior 1 [process 21278] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr:
------------------------------------------


------------------------------------------


---- [debuginfo-gdb] debuginfo/pretty-std-collections.rs stdout ----
NOTE: compiletest thinks it is using GDB with native rust support
NOTE: compiletest thinks it is using GDB version 8001001

error: line not found in debugger output: $1 = BTreeSet(size=15) = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14}
status: exit status: 0
command: "/usr/bin/gdb" "-quiet" "-batch" "-nx" "-command=/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo/pretty-std-collections.gdb/pretty-std-collections.debugger.script"
------------------------------------------
GNU gdb (Ubuntu 8.1.1-0ubuntu1) 8.1.1
Copyright (C) 2018 Free Software Foundation, Inc.
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Breakpoint 1 at 0x37d26: file /checkout/src/test/debuginfo/pretty-std-collections.rs, line 158.
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Breakpoint 1, pretty_std_collections::main () at /checkout/src/test/debuginfo/pretty-std-collections.rs:158
158     zzz(); // #break
$1 = BTreeSet(size=15)
$2 = BTreeSet(size=0)
$3 = BTreeMap(size=15) = {[0] = 0, [1] = 1, [2] = 2, [3] = 3, [4] = 4, [5] = 5, [6] = 6, [7] = 7, [8] = 8, [9] = 9, [10] = 10, [11] = 11, [12] = 12, [13] = 13, [14] = 14}
$4 = BTreeMap(size=0)
$5 = BTreeMap(size=2) = {[false] = core::option::Option<MaybeUninit<core::option::Option<bool>, [true] = core::option::Option<MaybeUninit<core::option::Option<bool>}
$6 = BTreeMap(size=15) = {[0] = pretty_std_collections::MyLeafNode (0), [1] = pretty_std_collections::MyLeafNode (1), [2] = pretty_std_collections::MyLeafNode (2), [3] = pretty_std_collections::MyLeafNode (3), [4] = pretty_std_collections::MyLeafNode (4), [5] = pretty_std_collections::MyLeafNode (5), [6] = pretty_std_collections::MyLeafNode (6), [7] = pretty_std_collections::MyLeafNode (7), [8] = pretty_std_collections::MyLeafNode (8), [9] = pretty_std_collections::MyLeafNode (9), [10] = pretty_std_collections::MyLeafNode (10), [11] = pretty_std_collections::MyLeafNode (11), [12] = pretty_std_collections::MyLeafNode (12), [13] = pretty_std_collections::MyLeafNode (13), [14] = pretty_std_collections::MyLeafNode (14)}
$7 = BTreeMap(size=1) = {[()] = 1}
$8 = BTreeMap(size=1)
$9 = BTreeMap(size=1) = {[()] = ()}
$10 = VecDeque(size=3) = {5, 3, 7}
$11 = VecDeque(size=7) = {2, 3, 4, 5, 6, 7, 8}
$12 = HashMap(size=4) = {[1] = 10, [2] = 20, [3] = 30, [4] = 40}
$13 = HashSet(size=4) = {1, 2, 3, 4}
A debugging session is active.

 Inferior 1 [process 21300] will be killed.

Quit anyway? (y or n) [answered Y; input not from terminal]
------------------------------------------
stderr:
------------------------------------------
------------------------------------------
Python Exception <class 'gdb.error'> No type named alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, MaybeUninit<i32, Result<(), alloc::collections::btree::node::marker::LeafOrInternal>.: 
Python Exception <class 'gdb.error'> No type named alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Owned, MaybeUninit<i32, Result<(), alloc::collections::btree::node::marker::LeafOrInternal>.: 
------------------------------------------



---
test result: FAILED. 84 passed; 5 failed; 45 ignored; 0 measured; 0 filtered out; finished in 3.75s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/debuginfo" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/debuginfo" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "debuginfo" "--mode" "debuginfo" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-10/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "10.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine frontendopenmp fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcerror orcjit passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvutils runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86utils xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:15:05
```

</details>

Sl1mb0 commented:
It's interesting that `pretty-slices.rs` was the only test that failed for me locally, but it's not the only test that failed for
`(x86_64-gnu-llvm-10, ubuntu-latest-xl)`.

michaelwoerister commented on compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs:
```diff
@@ -50,7 +52,14 @@ fn push_debuginfo_type_name<'tcx>(
+    type_name_cache: &RefCell<FxHashMap<(Ty<'tcx>, bool), String>>,
 ) {
+    // Check if we have seen this type and qualifier before.
+    if let Some(type_name) = type_name_cache.borrow().get(&(&t, qualified)) {
```
Does `.get(&(t, qualified))` work too?

michaelwoerister commented on compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs:
```diff
@@ -50,7 +52,14 @@ fn push_debuginfo_type_name<'tcx>(
 ) {
+    // Check if we have seen this type and qualifier before.
+    if let Some(type_name) = type_name_cache.borrow().get(&(&t, qualified)) {
+        output.push_str(&type_name.clone());
```
This should be `output.push_str(&type_name[..]);`. No need to clone the string for accessing its contents.

michaelwoerister commented on compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs:
```diff
@@ -375,6 +424,10 @@ fn push_debuginfo_type_name<'tcx>(
 
+    if type_name_cache.borrow_mut().insert((&t, qualified), output.clone()).is_some() {
+        bug!("type name is already in the type name cache!");
+    }
```
I think this is in the wrong place. `push_debuginfo_typename()` can be called recursively, so adding the contents here might insert half-built names into the cache. It looks like a likely cause for the test failures we are seeing. Try to move this to the end of `compute_debuginfo_type_name()` where we know that we are always dealing with complete type names.

michaelwoerister commented on compiler/rustc_codegen_ssa/src/debuginfo/type_names.rs:
```diff
@@ -33,12 +34,13 @@ pub fn compute_debuginfo_type_name<'tcx>(
     tcx: TyCtxt<'tcx>,
     t: Ty<'tcx>,
     qualified: bool,
+    type_name_cache: &RefCell<FxHashMap<(Ty<'tcx>, bool), String>>,
```
Can you try to make this `&mut FxHashMap<(Ty<'tcx>, bool), String>` instead of `&RefCell<...>` everywhere in this file and then do 
```rust
compute_debuginfo_type_name(cx.tcx, ty, qualified, debug_context(cx).type_name_cache.borrow_mut())
```
in `check_type_name_cache()`? It should be sufficient to secure mutable access to the cache once in the beginning.

Sl1mb0 replied:
So I tried this, but `borrow_mut()` returns a `MappedWriteGuard`. This works if we use `&mut `debug_context(cx)..` though.

michaelwoerister replied:
Yes, that's what I meant. I wasn't sure if the compiler would do that automatically or if the explicit &mut would be needed.

michaelwoerister reviewed (commented):
I left a few comments and might have found the reason for test failures. Give it a try `:)`

michaelwoerister commented on compiler/rustc_codegen_llvm/src/debuginfo/mod.rs:
```diff
@@ -438,12 +438,11 @@ impl DebugInfoMethods<'tcx> for CodegenCx<'ll, 'tcx> {
             substs: SubstsRef<'tcx>,
             name_to_append_suffix_to: &mut String,
         ) -> &'ll DIArray {
-            let type_name_cache = &mut *debug_context(cx).type_name_cache.borrow_mut();
```
I think holding the RefCell "lock" here for longer than just the `type_names::push_generic_params()` call might have been the reason why you saw the BorrowMut panics because the `type_metadata()` call farther below will also try to lock the cache.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3691094931) [(plain)](https://github.com/rust-lang/rust/commit/75886599c98934b9ea068da0762e79ed3a000c5e/checks/3691094931/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs at line 375:
 }
 
 fn check_type_name_cache(cx: &CodegenCx<'ll, 'tcx>, ty: Ty<'tcx>, qualified: bool) -> String {
-    compute_debuginfo_type_name(cx.tcx, ty, qualified, &mut debug_context(cx).type_name_cache.borrow_mut())
+    compute_debuginfo_type_name(
+        cx.tcx,
+        ty,
+        qualified,
+        &mut debug_context(cx).type_name_cache.borrow_mut(),
 }
 
 fn fixed_vec_metadata(
 fn fixed_vec_metadata(
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_llvm/src/back/write.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/utils.rs" "/checkout/compiler/rustc_codegen_llvm/src/back/profiling.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/diagnostic.rs" "/checkout/compiler/rustc_codegen_llvm/src/abi.rs" "/checkout/compiler/rustc_codegen_llvm/src/llvm/archive_ro.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/metadata.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

michaelwoerister commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit fdd211a5d12344013061ba97da23a65cdca53c33 with merge 7cd33b678f6735d7b26fb78172256d815cb27b33...
<!-- homu: {"type":"TryBuildStarted","head_sha":"fdd211a5d12344013061ba97da23a65cdca53c33","merge_sha":"7cd33b678f6735d7b26fb78172256d815cb27b33"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3692805646)
Build commit: 7cd33b678f6735d7b26fb78172256d815cb27b33 (`7cd33b678f6735d7b26fb78172256d815cb27b33`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3692805646"},"merge_sha":"7cd33b678f6735d7b26fb78172256d815cb27b33"} -->

rust-timer commented:
Queued 7cd33b678f6735d7b26fb78172256d815cb27b33 with parent bf642323d621dcefeef1d8ab4711aae36e357615, future [comparison URL](https://perf.rust-lang.org/compare.html?start=bf642323d621dcefeef1d8ab4711aae36e357615&end=7cd33b678f6735d7b26fb78172256d815cb27b33).

rust-timer commented:
Finished benchmarking commit (7cd33b678f6735d7b26fb78172256d815cb27b33): [comparison url](https://perf.rust-lang.org/compare.html?start=bf642323d621dcefeef1d8ab4711aae36e357615&end=7cd33b678f6735d7b26fb78172256d815cb27b33).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf. 

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

michaelwoerister reviewed (commented):
It looks like this doesn't improve performance much so far. We can try to improve things a little more by avoiding some string data allocations as described below. But I think that's as far as we can get as long as we don't find another way of dealing with recursive types.

Sl1mb0 commented:
Do you think it's worth exploring recording how many times we either access the cache or compute the string?

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/3794833458) [(plain)](https://github.com/rust-lang/rust/commit/13a502d2e8380c4baaf4bc67bf3d67fa03635567/checks/3794833458/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
Diff in /checkout/compiler/rustc_codegen_llvm/src/debuginfo/mod.rs at line 38:
 use libc::c_uint;
 use smallvec::SmallVec;
 use std::cell::RefCell;
-use std::rc::Rc;
 use std::iter;
+use std::rc::Rc;
 use tracing::debug;
 
 mod create_scope_map;
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2018" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/utils.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/gdb.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/mod.rs" "/checkout/compiler/rustc_save_analysis/src/lib.rs" "/checkout/compiler/rustc_save_analysis/src/dumper.rs" "/checkout/compiler/rustc_codegen_llvm/src/debuginfo/create_scope_map.rs" "/checkout/compiler/rustc_incremental/src/assert_module_sources.rs" "/checkout/compiler/rustc_save_analysis/src/sig.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

nagisa commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 353e42e8a3d79173c4ea91395c54624413d623c8 with merge a35f2fa0b1709e3bffed26d72fff54832568660f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"353e42e8a3d79173c4ea91395c54624413d623c8","merge_sha":"a35f2fa0b1709e3bffed26d72fff54832568660f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/3797596205)
Build commit: a35f2fa0b1709e3bffed26d72fff54832568660f (`a35f2fa0b1709e3bffed26d72fff54832568660f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/3797596205"},"merge_sha":"a35f2fa0b1709e3bffed26d72fff54832568660f"} -->

rust-timer commented:
Queued a35f2fa0b1709e3bffed26d72fff54832568660f with parent 175b8db73bfd078b4bcd3c28c8d6f51d5895ebf3, future [comparison URL](https://perf.rust-lang.org/compare.html?start=175b8db73bfd078b4bcd3c28c8d6f51d5895ebf3&end=a35f2fa0b1709e3bffed26d72fff54832568660f).

rust-timer commented:
Finished benchmarking commit (a35f2fa0b1709e3bffed26d72fff54832568660f): [comparison url](https://perf.rust-lang.org/compare.html?start=175b8db73bfd078b4bcd3c28c8d6f51d5895ebf3&end=a35f2fa0b1709e3bffed26d72fff54832568660f).

**Summary**: This benchmark run did not return any relevant changes.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf. 

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf -perf-regression

michaelwoerister commented:
Well, it looks like there are no real performance improvements to be had here, unfortunately. Given that the changes add some complexity with no apparent gain, I'd opt for closing the PR (unless you have other ideas that we haven't explored yet).

michaelwoerister commented:
I hope you still had an interesting learning experience working on this, @Sl1mb0!
